sw/device/lib/dif/dif_padctrl.h

```diff
@@ -0,0 +1,244 @@
+
+#include <stdbool.h>
+
+// Generated.
+#include "padctrl_regs.h"
+
+#include "sw/device/lib/base/mmio.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Pad attributes.
+ *
+ * Some attributes have defined meanings, others are implementation defined. Not
+ * all pads will support all attributes. The padctrl DIF does not check for
+ * unsupported attributes. To check to see if an attribute is supported after
+ * enabling it (via `dif_padctrl_attr_enable` or `dif_padctrl_attr_write`) read
+ * back the attribute field (via `dif_padctrl_attr_get`). If the attribute isn't
+ * supported the attribute will still be disabled (relevant bit unset in the
+ * attribute field).
+ */
+typedef enum dif_padctrl_attr {
+  kDifPadctrlAttrNone = 0,
+  kDifPadctrlAttrIOInvert = 1 << 0,
+  kDifPadctrlAttrOpenDrain = 1 << 2,
+  kDifPadctrlAttrPullDown = 1 << 3,
+  kDifPadctrlAttrPullUp = 1 << 4,
+  kDifPadctrlAttrKeeper = 1 << 5,
+  kDifPadctrlAttrWeakDrive = 1 << 6,
+} dif_padctrl_attr_t;
+
+/**
+ * Generic padctrl DIF return codes
+ */
+typedef enum dif_padctrl_result {
+  /**
+   * Operation succeeded.
+   */
+  kDifPadctrlOk = 0,
+  /**
+   * Operation failed due to an unspecified error. It cannot be recovered or
+   * retried. */
+  kDifPadctrlError = 1,
+  /**
+   * Operation failed due to a bad argument and can be retried. No register
+   * reads or writes have been performed.
+   */
+  kDifPadctrlBadArg = 2,
+} dif_padctrl_result_t;
+
+/**
+ * Pad kinds.
+ *
+ * A pad can be an MIO or DIO pad
+ */
+typedef enum dif_padctrl_pad_kind {
+  /**
+   * MIO (Multiplexed Input/Output) pad. Connected via pinmux.
+   */
+  kDifPadctrlPadMio = 0,
+  /**
+   * DIO (Dedicated Input/Output) pad. Connected directly to peripheral inputs
+   * and outputs, bypassing pinmux.
+   */
+  kDifPadctrlPadDio,
+} dif_padctrl_pad_kind_t;
+
+/**
+ * Pad index.
+ *
+ * Combined with `dif_padctrl_pad_kind_t` to refer to a specific pad.
+ */
+typedef uint32_t dif_padctrl_pad_index_t;
+
+/**
+ * Padctrl instance state.
+ *
+ * Padctrl persistent data that is required by the Padctrl API.
+ */
+typedef struct dif_padctrl {
+  mmio_region_t base_addr; /**< Padctrl base address. */
+} dif_padctrl_t;
+
+/**
+ * Initialises an instance of Padctrl.
+ *
+ * Information that must be retained, and is required to program Padctrl, shall
+ * be stored in @p padctrl.
+ *
+ * @param base_addr Base address of an isntance of the Padctrl IP block
+ * @param padctrl Padctrl state data
+ * @return dif_padctrl_init_result_t.
+ */
+dif_padctrl_result_t dif_padctrl_init(mmio_region_t base_addr,
+                                      dif_padctrl_t *padctrl);
+
+/**
+ * Padctrl reset routine error codes.
+ */
+typedef enum dif_padctrl_reset_result {
+  kDifPadctrlResetOk = kDifPadctrlOk,
+  kDifPadctrlResetError = kDifPadctrlError,
+  kDifPadctrlResetBadArg = kDifPadctrlBadArg,
+  /**
+   * Peripheral is locked and cannot be reset.This is not recoverable.
```

Comment:
Missed a space after the period.

Though I would put "This is not recoverable." on a line of its own.